src/backends/plonky2/recursion/circuit.rs

```diff
@@ -0,0 +1,627 @@
+/// This file contains the RecursiveCircuit, the circuit used for recursion.
+///
+/// The RecursiveCircuit verifies N proofs of itself (N=arity), together with
+/// the logic defined at the InnerCircuit (in our case, used for the
+/// MainPodCircuit logic).
+///
+/// The arity defines the maximum amount of proofs of itself that the
+/// RecursiveCircuit verifies. When arity>1, using the RecursiveCircuit has the
+/// shape of a tree of the same arity.
+///                                              
+//
+//                      π_root
+//                        ▲
+//                ┌───────┴────────┐
+//                │RecursiveCircuit│
+//                └─▲───▲───▲────▲─┘
+//          ┌───────┘  ┌┘   └┐   └──────┐
+//          │π''_1     │ ... │     π''_N│
+// ┌────────┴───────┐ ┌┴┐┌─┐┌┴┐ ┌───────┴────────┐
+// │RecursiveCircuit│ │.││.││.│ │RecursiveCircuit│
+// └──▲─────────▲───┘ └─┘└─┘└─┘ └──▲─────────▲───┘
+//    │         │                  │         │
+//   π_1  ...  π_N               π'_1 ...  π'_N
+//
+//
+// where
+// N: arity of the RecursiveCircuit
+// π_i: plonky2 proof of the RecursiveCircuit
+//
+use hashbrown::HashMap;
+use plonky2::{
+    self,
+    gates::noop::NoopGate,
+    iop::{
+        target::{BoolTarget, Target},
+        witness::{PartialWitness, WitnessWrite},
+    },
+    plonk::{
+        circuit_builder::CircuitBuilder,
+        circuit_data::{
+            CircuitConfig, CircuitData, ProverCircuitData, VerifierCircuitData,
+            VerifierCircuitTarget,
+        },
+        proof::{ProofWithPublicInputs, ProofWithPublicInputsTarget},
+    },
+    recursion::dummy_circuit::cyclic_base_proof,
+};
+
+use crate::{
+    backends::plonky2::{
+        basetypes::{Proof, C, D},
+        error::{Error, Result},
+    },
+    middleware::F,
+};
+
+/// InnerCircuit is the trait used to define the logic of the circuit that is
+/// computed inside the RecursiveCircuit.
+pub trait InnerCircuit: Clone {
+    type Input;
+
+    fn build(
```

Comment:
what about parametrizing the circuit?
Can we add a `params: &Self::Params` to this method?
Then `RecursiveCircuit::build` could take `inner_params: &I::Params`

It would look like this:

```rust
pub trait InnerCircuit: Clone {
    type Input;
    type Params;

    fn build(
        builder: &mut CircuitBuilder<F, D>,
        params: &Self::Params,
        // public_inputs will be used for the recursive proof verification
        public_inputs: Vec<Target>,
        selectors: Vec<BoolTarget>,
    ) -> Result<Self>;
```

Reply:
Right, haven't added it before bcs didn't want to guess future needs of the interface (that will get clear once the RecursiveCircuit & InnerCircuit start to get used in the rest of the Pod codebase), but this one seems a clear thing that will be used; added it at the commit https://github.com/0xPARC/pod2/commit/794eaa8612b606f64ed164a2cf5d2dd16f3ae10e